src/app/(app)/product-list.tsx

```diff
@@ -1,95 +1,191 @@
+    if (products.length === 0) {
+      return (
+        <View className="m-4 gap-4 self-center text-4xl font-semibold">
+          <Text>No products found</Text>
         </View>
-      </SafeAreaView>
-    </>
+      );
+    }
+
+    return (
+      <View className="mt-4">
+        <FlatList
+          onEndReached={onEndReached}
+          data={products}
+          onScroll={onScroll}
+          renderItem={({ item, index }) => {
+            const isFirstItem = index === 0;
+            const isLastItem = index === products.length - 1;
+
+            return (
+              <View
+                className={`mx-4 border ${isFirstItem ? 'rounded-t-lg' : ''} ${
+                  isLastItem ? 'rounded-b-lg' : ''
+                }`}
+              >
+                <ProductCard
+                  id={item.id}
+                  price={item.unit_price}
+                  state={item.state}
+                  name={item.title}
+                  image_url={item.pictures[0]}
+                />
+              </View>
+            );
+          }}
+          keyExtractor={(item) => item.id.toString()}
+        />
+      </View>
+    );
+  }
+);
+
+const useProductPagination = (
+  page: number,
+  data: FetchProductsResponse | undefined,
+  setPage: React.Dispatch<React.SetStateAction<number>>
+) => {
+  const handleEndReached = useCallback(() => {
+    if (page * 7 - 1 < (data?.pagination?.count ?? 0)) {
+      setPage((prev) => prev + 1);
+    }
+  }, [page, data, setPage]);
+
+  const handleScroll = useCallback(
+    ({ nativeEvent }: any) => {
+      const { layoutMeasurement, contentOffset, contentSize } = nativeEvent;
+      if (
+        layoutMeasurement.height + contentOffset.y >=
+        contentSize.height - 20
+      ) {
+        handleEndReached();
+      }
+    },
+    [handleEndReached]
+  );
+
+  return { handleEndReached, handleScroll };
+};
+
+export default function ProducList() {
+  const [page, setPage] = useState(1);
+  const [products, setProducts] = useState<Product[]>([]);
+  const { data } = useProducts({ variables: { page: page, items: 7 } });
+  const [fetchedProducts, setFetchedProducts] = useState<Product[]>([]);
+  const [query, setQuery] = useState<string>('');
+  const [shouldReset, setShouldReset] = useState(false);
+  const { handleEndReached, handleScroll } = useProductPagination(
+    page,
+    data,
+    setPage
+  );
+
+  const resetSearchBar = () => {
+    setProducts(fetchedProducts);
+    setShouldReset(true);
+    setQuery('');
+    setTimeout(() => setShouldReset(false), 100);
+  };
+
+  const handleProductSelect = (searchQuery: string) => {
+    if (searchQuery === '') {
+      resetSearchBar();
+      return;
+    }
+    const queryProducts = fetchedProducts.filter((product) =>
+      product.title.toLowerCase().includes(searchQuery.toLowerCase())
+    );
+    setQuery(searchQuery);
+    setProducts(queryProducts.slice(0, 7));
+  };
+
```

Comment:
From what I read in the API docs the products query has a text queryParam that should be used to search, that's the only way you can actually search through the entire database despite which pages you've already loaded.

check https://rs-blackmarket-api.herokuapp.com/api-docs/index.html

Reply:
FetchedProducts is to save the accumulation of fetched products without searching and Products is for the products that are showed on the list. I shouldn't keep them to switch the products displayed? Or is possible do that without the states?

Reply:
I think I might understand the confusion now, the same way you have 'createQuery' function, there's also one specific when using pagination `createInfiniteQuery` https://github.com/HuolalaTech/react-query-kit?tab=readme-ov-file#createinfinitequery.
Is in this function that the accumulated results of each page will live, so the data returned by this inifiniteQuery should contain all pages.

I might be missing something and that's why I want to hop on a call , but I think you just need to use the data returned by the query you want in a given moment, it might have a search term or it might not, I don't think it's necessary to be handling extra state values